src/channel/NodeChannel.js

```diff
@@ -18,73 +18,151 @@
+
+                            resolve(certPEM);
+                        } else {
+                            reject(new Error("No certificate retrieved."));
+                        }
+                    } catch (err) {
+                        reject(err);
+                    } finally {
+                        socket.end();
+                    }
+                },
+            );
+
+            socket.on("error", reject);
+        });
+    }
+
+    /**
+     * Initialize the gRPC client
+     * @returns {Promise<void>}
+     */
+    async _initializeClient() {
+        if (clientCache[this.address]) {
+            this._client = clientCache[this.address];
+            return;
+        }
+
         let security;
-        let options;
-
-        if (this.cert != null) {
-            security = credentials.createSsl(Buffer.from(this.cert));
-            options = {
-                "grpc.ssl_target_name_override": "127.0.0.1",
-                "grpc.default_authority": "127.0.0.1",
-                "grpc.http_connect_creds": "0",
-                // https://github.com/grpc/grpc-node/issues/1593
-                // https://github.com/grpc/grpc-node/issues/1545
-                // https://github.com/grpc/grpc/issues/13163
-                "grpc.keepalive_time_ms": 100000,
-                "grpc.keepalive_timeout_ms": 10000,
-                "grpc.keepalive_permit_without_calls": 1,
-                "grpc.enable_retries": 0,
-            };
+        const options = {
+            "grpc.ssl_target_name_override": "127.0.0.1",
+            "grpc.default_authority": "127.0.0.1",
+            "grpc.http_connect_creds": "0",
+            "grpc.keepalive_time_ms": 100000,
+            "grpc.keepalive_timeout_ms": 10000,
+            "grpc.keepalive_permit_without_calls": 1,
+            "grpc.enable_retries": 0,
+        };
+
+        // If the port is 50212, use TLS
+        if (this.nodePort === "50212") {
+            const certificate = Buffer.from(
+                await this._retrieveCertificate(this.address),
+            );
+
+            security = credentials.createSsl(certificate);
         } else {
             security = credentials.createInsecure();
-            options = {
-                "grpc.keepalive_time_ms": 100000,
-                "grpc.keepalive_timeout_ms": 10000,
-                "grpc.keepalive_permit_without_calls": 1,
-                "grpc.enable_retries": 0,
-            };
         }
 
-        /**
-         * @type {Client}
-         * @private
-         */
-        this._client = new Client(address, security, options);
+        this._client = new Client(this.address, security, options);
+
+        clientCache[this.address] = this._client;
     }
 
     /**
      * @override
      * @returns {void}
      */
     close() {
-        this._client.close();
+        if (this._client) {
+            this._client.close();
+            delete clientCache[this.address];
```

Comment:
Why not clean the cert cache as well?

Reply:
That's a good question, I might have forgotten about it. I'll make sure to clean the cert as well.